pkg/clients/integration/pipelineruns.go

```diff
@@ -180,6 +181,45 @@ func (i *IntegrationController) WaitForAllIntegrationPipelinesToBeFinished(testN
 	return nil
 }
 
+// WaitForFinalizerToGetRemovedFromAllIntegrationPipelineRuns waits for
+// the given finalizer to get removed from all integration pipelinesruns
+// that are related to the given application and namespace.
+func (i *IntegrationController) WaitForFinalizerToGetRemovedFromAllIntegrationPipelineRuns(testNamespace, applicationName string, snapshot *appstudioApi.Snapshot) error {
+	integrationTestScenarios, err := i.GetIntegrationTestScenarios(applicationName, testNamespace)
```

Comment:
I feel you may just get the PLRs with `application.Name` in their label since we add `application.Name` to label of integration test PLR when we create it in https://github.com/redhat-appstudio/integration-service/blob/main/tekton/integration_pipeline.go#L214, then we can save one query. How do you think of it?

Reply:
That's a good idea, I was halfway into implementing your suggestion but then I realised that even with this approach, we'll be making at least 2 calls to the cluster. For example:

Currently, we're doing this:
* Calling the cluster to fetch all the ITS associated with the given application.
* For each of these ITS, we're repeatedly calling the cluster to fetch the related Integration PLR which doesn't have a finalizer attached to it.

With your suggestion, we'll be doing this:
* Calling the cluster to fetch all the Integration PLRs associated with the given application.
* For each of these PLRs, we're repeatedly calling the cluster to fetch its latest version of Integration PLR which doesn't have a finalizer attached to it.

So, we would have to make at least 2 calls in both approaches. I think the current way of checking the PLRs is sufficient.

Reply:
> That's a good idea, I was halfway into implementing your suggestion but then I realised that even with this approach, we'll be making at least 2 calls to the cluster. For example:
> 
> Currently, we're doing this:
> 
>     * Calling the cluster to fetch all the ITS associated with the given application.
> 
>     * For each of these ITS, we're repeatedly calling the cluster to fetch the related Integration PLR which doesn't have a finalizer attached to it.
> 
> 
> With your suggestion, we'll be doing this:
> 
>     * Calling the cluster to fetch all the Integration PLRs associated with the given application.
> 
>     * For each of these PLRs, we're repeatedly calling the cluster to fetch its latest version of Integration PLR which doesn't have a finalizer attached to it.

Why do you need to fetch its latest version of integration PLR? Ideally, the finalizer should be removed from all integration PLR, not only the latest version of PLR.

Reply:
Because if we don't fetch the latest copy of the PLR, this check: https://github.com/redhat-appstudio/e2e-tests/blob/c3fa2ab8e13fb7de1ad09a1b8314c80631fcf57b/pkg/clients/integration/pipelineruns.go#L214

would always return true.

Because the `pipelineRun` variable here would contain the finalizer in it. And after a few seconds, the integration service would remove the finalizer from this PLR but the above `pipelineRun` variable would still contain the older/outdated PLR that still contains the finalizer. This is why we need to constantly update this variable to contain the latest version of the PLR, on this line: https://github.com/redhat-appstudio/e2e-tests/blob/c3fa2ab8e13fb7de1ad09a1b8314c80631fcf57b/pkg/clients/integration/pipelineruns.go#L209

Reply:
I think once one snapshot has passed test and status is reported to snapshot status annotation, the finalizer should be removed from all PLRs created for this snapshot, otherwise, it might be a bug.

Reply:
@hongweiliu17 that's correct, but the object that's loaded in the `pipelineRun` variable won't contain the latest state of the custom resource on the cluster, hence the need to query the cluster again to get that latest state which should have the finalizer removed from it.